### PR TITLE
Add support for SapMachine

### PR DESCRIPTION
**Pull request by joschi**
SapMachine is an OpenJDK release maintained and supported by SAP.

https://sap.github.io/SapMachine/

### CODE REVIEW

File: bin/functions [hunk 1]
```diff
@@ -1,7 +1,9 @@
 #!/usr/bin/env bash
+set -e
+set -o pipefail
```
Comment (joschi) [inline bin/functions]: The script should fail if something goes wrong.

File: bin/functions [hunk 2]
```diff
@@ -1,7 +1,9 @@
...
 
 PLUGIN_HOME=$(dirname $(dirname "${0}"))
-CACHE_DIR="/tmp/asdf-java.cache"
+CACHE_DIR="${TMPDIR:-/tmp/}asdf-java.cache"
```
Comment (joschi) [inline bin/functions]: Use the user-specific temporary directory if available.

**Comment by halcyon** [comment]
@joschi Are these failures specific to travis? Is travis redundant now that we are using github actions?

**Comment by joschi** [comment]
I'm not sure about the build failures of this PR.
It works on my machine with macOS and in a Docker container with Ubuntu Linux.

This being said, I'll check this PR again over the weekend, so let's not merge it for now.

And yes, I think Travis CI build jobs could be removed in favor of GitHub actions.

**Comment by joschi** [comment]
@halcyon Sorry for the long delay. This PR is working now as expected. 😃

**Comment by halcyon** [comment]
```
% rm -rf /tmp/asdf-java.cache  
% asdf list-all java 2>&1
parse error: Invalid numeric literal at line 1, column 4
adopt-openjdk-10.0.2+13                                                  
adopt-openjdk-10.0.2+13.1
...
```

File: bin/functions [hunk 3]
```diff
@@ -44,29 +44,62 @@ function check-jq() {
...
-            mv ${CACHE_DIR}/${i}.temp ${CACHE_DIR}/${i}
+               ${i} > ${i}.temp
+            mv ${i}.temp ${i}
         done
     fi
 }
 
+function retrieve-sapmachine() {
+    args=('-s' '-f')
+    if [ -n "$GITHUB_API_TOKEN" ]; then
```
Comment (halcyon) [inline bin/functions]:
```suggestion
    if [ -n "${GITHUB_API_TOKEN}" ]; then
```

File: bin/functions [hunk 4]
```diff
@@ -44,29 +44,62 @@ function check-jq() {
...
+        args+=('-H' "Authorization: token $GITHUB_API_TOKEN")
```
Comment (halcyon) [inline bin/functions]:
```suggestion
        args+=('-H' "Authorization: token ${GITHUB_API_TOKEN}")
```

File: bin/functions (hunk 1 above)
Comment (halcyon) [reply bin/functions]: https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo_pipefail/

**Comment by joschi** [comment]
> Would be nice if retrieve-sapmachine() was generic and supported retrieving Github Releases from other Github repos. 

@halcyon I thought the same. Either that or "caching" the result for all JVM flavors with a scheduled GitHub action (cron job) in this or in a separate repository, so that we can reduce the number of remote requests on the clients.

What do you think about the latter? Would it make sense to implement that?

File: bin/functions (hunk 1 above)
Comment (joschi) [reply bin/functions]: Yes, unfortunately this didn't work out for now.
I would try to address this in a separate PR.

**Comment by halcyon** [comment]
> > Would be nice if retrieve-sapmachine() was generic and supported retrieving Github Releases from other Github repos.
> 
> @halcyon I thought the same. Either that or "caching" the result for all JVM flavors with a scheduled GitHub action (cron job) in this or in a separate repository, so that we can reduce the number of remote requests on the clients.
> 
> What do you think about the latter? Would it make sense to implement that?

That's an interesting idea - I'll have to think about it some more, but I think I initially favor caching on the client side

**Comment by halcyon** [comment]
Thank you!

**Pull request merged**